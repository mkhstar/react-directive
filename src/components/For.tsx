import React from "react";
import { ForProps } from "../types/for";

const For = <T,>({ each, children }: ForProps<T>) => (
  <>{each.map((value, index) => children(value, index))}</>
);

export default For;
