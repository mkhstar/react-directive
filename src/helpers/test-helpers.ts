import { ReactElement } from "react";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";

export const snapshotTest = (
  render: ReactElement<any, string>,
  title?: string
) => {
  const tree = renderer.create(render).toJSON();
  expect(tree).toMatchSnapshot(title || "");
};

export const truthyValues = ["value", true, 1, Infinity, -1, -Infinity, {}, []];
export const falsyValues = ["", false, 0, null];
export const truthyFalsyMap = truthyValues
  .map((value) =>
    falsyValues.map((falsyValue) => ({ truthy: value, falsy: falsyValue }))
  )
  .flat();
