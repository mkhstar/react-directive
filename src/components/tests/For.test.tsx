import React from "react";
import { render } from "@testing-library/react";
import { snapshotTest } from "../../helpers/test-helpers";
import { For } from "../For";

describe("<For />", () => {
  it("renders correct number of children", () => {
    const each = [1, 2, 3];
    const { queryAllByTestId } = render(
      <For each={each}>
        {(value) => (
          <div key={value} data-testid="child">
            {value}
          </div>
        )}
      </For>
    );

    expect(queryAllByTestId("child").length).toBe(each.length);
  });

  it("renders children with correct values", () => {
    const each = [1, 2, 3];
    const { queryAllByText } = render(
      <For each={each}>{(value) => <div key={value}>{value}</div>}</For>
    );

    each.forEach((value) => {
      expect(queryAllByText(value.toString()).length).toBe(1);
    });
  });

  it("does not render anything if `each` is an empty array", () => {
    const each: Array<number> = [];
    const { queryAllByTestId } = render(
      <For each={each}>{(value) => <div key={value}>{value}</div>}</For>
    );

    expect(queryAllByTestId("child").length).toBe(0);
  });

  it("passes the correct value and index to children", () => {
    const each = [
      { id: 1, name: "Musah", school: "Hacettepe", from: "Kumasi" },
      { id: 2, name: "Shakino", school: "Okess", from: "Oman" },
      { id: 3, name: "Sala", school: "UOEW", from: "Kumasi" },
    ];
    const { queryAllByText } = render(
      <For each={each}>
        {(value, index) => (
          <div
            key={value.id}
          >{`Name: ${value.name}, School ${value.school}, From ${value.from}, Index: ${index}`}</div>
        )}
      </For>
    );

    each.forEach((value, index) => {
      expect(
        queryAllByText(
          `Name: ${value.name}, School ${value.school}, From ${value.from}, Index: ${index}`
        ).length
      ).toBe(1);
    });
  });

  it("passes the correct value and index to children with multiple nodes", () => {
    const each = [{ name: "Musah", school: "Hacettepe", from: "Kumasi" }];
    const { queryByTestId } = render(
      <For each={each}>
        {(value, index) => (
          <React.Fragment key={value.from}>
            <p data-testid="name">Name: {value.name}</p>
            <p data-testid="school">School: {value.school}</p>
            <p data-testid="from">From: {value.from}</p>
            <p data-testid="index">Index: {index}</p>
          </React.Fragment>
        )}
      </For>
    );
    expect(queryByTestId("name")?.textContent).toEqual("Name: Musah");
    expect(queryByTestId("school")?.textContent).toEqual("School: Hacettepe");
    expect(queryByTestId("from")?.textContent).toEqual("From: Kumasi");
    expect(queryByTestId("index")?.textContent).toEqual("Index: 0");
  });
});

describe("<For /> snapshots", () => {
  it("matches the snapshot with an array of numbers", () => {
    snapshotTest(
      <For each={[1, 2, 3]}>
        {(value, index) => <div key={index}>{value}</div>}
      </For>
    );
  });

  it("matches the snapshot with an array of strings", () => {
    snapshotTest(
      <For each={["a", "b", "c"]}>
        {(value, index) => <div key={index}>{value}</div>}
      </For>
    );
  });

  it("matches the snapshot with an empty array", () => {
    snapshotTest(
      <For each={[]}>{(value, index) => <div key={index}>{value}</div>}</For>
    );
  });

  it("passes the correct value and index to children", () => {
    const each = [
      { id: 1, name: "Musah", school: "Hacettepe", from: "Kumasi" },
      { id: 2, name: "Shakino", school: "Okess", from: "Oman" },
      { id: 3, name: "Sala", school: "UOEW", from: "Kumasi" },
    ];

    snapshotTest(
      <For each={each}>
        {(value, index) => (
          <div
            key={value.id}
          >{`Name: ${value.name}, School ${value.school}, From ${value.from}, Index: ${index}`}</div>
        )}
      </For>
    );
  });
});
