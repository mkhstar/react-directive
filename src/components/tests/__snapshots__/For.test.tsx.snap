// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<For /> snapshots matches the snapshot with an array of numbers 1`] = `
[
  <div>
    1
  </div>,
  <div>
    2
  </div>,
  <div>
    3
  </div>,
]
`;

exports[`<For /> snapshots matches the snapshot with an array of strings 1`] = `
[
  <div>
    a
  </div>,
  <div>
    b
  </div>,
  <div>
    c
  </div>,
]
`;

exports[`<For /> snapshots matches the snapshot with an empty array 1`] = `null`;

exports[`<For /> snapshots passes the correct value and index to children 1`] = `
[
  <div>
    Name: Musah, School Hacettepe, From Kumasi, Index: 0
  </div>,
  <div>
    Name: Shakino, School Okess, From Oman, Index: 1
  </div>,
  <div>
    Name: Sala, School UOEW, From Kumasi, Index: 2
  </div>,
]
`;
